numba/tests/__init__.py

```diff
@@ -82,22 +92,59 @@ def _getParentArgParser(self):
         parser.add_argument('-m', '--multiprocess', dest='multiprocess',
                             action='store_true',
                             help='Parallelize tests')
+        parser.add_argument('-l', '--list', dest='list',
+                            action='store_true',
+                            help='List tests without running them')
         parser.add_argument('--profile', dest='profile',
                             action='store_true',
                             help='Profile the test run')
         return parser
 
     def parseArgs(self, argv):
+        if '-l' in argv:
+            argv.remove('-l')
+            self.list = True
         if PYVERSION < (3, 4) and '-m' in argv:
             # We want '-m' to work on all versions, emulate this option.
             argv.remove('-m')
             self.multiprocess = True
         super(NumbaTestProgram, self).parseArgs(argv)
+        # in Python 2.7, the 'discover' option isn't implicit.
+        # So if no test names were provided and 'self.test' is empty,
+        # we assume discovery hasn't been done yet.
+        if (not getattr(self, 'testNames', None) and
+            not self.test or (isinstance(self.test, suite.BaseTestSuite) and
+                              not self.test.countTestCases())):
+            self._do_discovery([])
+            
         if self.verbosity <= 0:
             # We aren't interested in informational messages / warnings when
             # running with '-q'.
             self.buffer = True
-
+            
+    def _do_discovery(self, argv, Loader=None):
+        """The discovery process is complicated by the fact that:
+
+        * different test suites live under different directories
+        * some test suites may not be available (CUDA)
+        * some tests may have to be run serially, even in the presence of the '-m' flag."""
+
+        from numba import cuda
+        loader = unittest.TestLoader()
+        topdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../..'))
+        base_tests = loader.discover('%s/numba/tests'%topdir, 'test*.py', topdir)
```

Comment:
It should be quite portable indeed, so `os.path.join` is perhaps mostly a matter of cleanliness or purity :)